guardian/atoms-rendering
# Fix `YoutubeAtom` race conditions and refactor

## What does this change?

This is a refactor of `YoutubeAtom` to address the following issues:

- multiple race conditions leading to:
  - events not sent to Ophan or GA (fixed in #330 )
  - the player not auto playing or sending events in the branch: https://github.com/guardian/dotcom-rendering/pull/3887 (fixed in this PR)
- using an iFrame embed _and_ the library `youtube-player` to wrap calls to the underlying YouTube player
- over time this has become an unwieldy component that mixes concerns between `Player`, `Overlay` and `Placeholder`
- internal component state is very difficult to reason about leading to brittle code and bugs

The new version of `YoutubeAtom` has the following changes:

- decoupled UI and player
- introduces a parent component `YoutubeAtom` and child components `YoutubeAtomPlayer`, `YoutubeAtomOverlay` and `YoutubeAtomPlaceholder`
- `YoutubeAtom` coordinates message passing between sibling components
- the player no longer inserts an iFrame embed
- the player only uses the library `youtube-player` to create the player and queue up api calls
- the player now has an `onReady` listener to properly determine when the player is ready
- the player `onStateChange` and `onReady` listeners use the provided YT player passed in the event object
- simplifies `Overlay` and `Placeholder` show/hide logic
- fixes `Placeholder` height and width issues
- improves `Placeholder` 'hide' behaviour  to wait for the player to be ready
- logging to allow inspection of events and component flow via `@guardian/libs logger` e.g. `guardian.logger.subscribeTo('dotcom')`

## How to test

Run a linked version on DCR with the following scenarios:

- a video with an overlay e.g. [link](https://www.theguardian.com/uk-news/2020/dec/04/edinburgh-hit-by-thundersnow-as-sonic-boom-wakes-residents)
- a video without an overlay i.e. it should show a placeholder that is removed when the player is ready
- an article with multiple atom videos e.g. [link](https://www.theguardian.com/media/2021/jun/25/guardian-journalists-win-orwell-prize-for-video-series-anywhere-but-westminster-john-harris-john-domokos)
- turn on logging to see the events

## Before and After

### Before

Islands branch, player does not trigger play or send Ophan events:

https://user-images.githubusercontent.com/7014230/154463150-53cf5852-28b7-49f4-a72d-02de6e7d3940.mp4

Placeholder, not sizing correctly and hiding too soon:

https://user-images.githubusercontent.com/7014230/154464217-c9a2ab69-3922-4b27-aa86-d74c8810bbd7.mp4

### After

Islands branch, player triggers play and sends Ophan events:

https://user-images.githubusercontent.com/7014230/154464441-31178949-f8e1-4b09-b2a3-9f7a89c0f748.mp4

Placeholder, sizing correctly and hiding when the player is ready:

https://user-images.githubusercontent.com/7014230/154464788-ddbe6b0b-edbf-427e-bfee-6dacbe67e561.mp4

Logging:

<img width="1171" alt="youtube-atom-logging-2" src="https://user-images.githubusercontent.com/7014230/154929938-8e2a775e-f8e1-4d93-87e9-5977a4bca775.png">

## Todo

Not necessarily for this PR but:

- improve the unit test coverage where possible - however I think e2e tests are better suited to player tests
- types could be improved to only use `@types/youtube` types
